From 21ef03b50ef64d25a05d7ac047c0e382237c9b15 Mon Sep 17 00:00:00 2001
From: Miklos Vajna <vmiklos@collabora.co.uk>
Date: Tue, 19 Jun 2018 15:45:42 +0000
Subject: [PATCH] Add FPDFText_GetMatrix() API
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This is similar to FPDFPath_GetMatrix(), but works on text, not path
objects.

Change-Id: If268362b7fa4398124b953e0e2225074523f5f65
Reviewed-on: https://pdfium-review.googlesource.com/35434
Reviewed-by: dsinclair <dsinclair@chromium.org>
Reviewed-by: Nicol치s Pe침a Moreno <npm@chromium.org>
Commit-Queue: Nicol치s Pe침a Moreno <npm@chromium.org>
---
 fpdfsdk/fpdf_edit_embeddertest.cpp | 17 +++++++++++++++++
 fpdfsdk/fpdf_edittext.cpp          | 30 ++++++++++++++++++++++++++++++
 public/fpdf_edit.h                 | 25 +++++++++++++++++++++++++
 3 files changed, 72 insertions(+)

diff --git a/fpdfsdk/fpdf_edittext.cpp b/fpdfsdk/fpdf_edittext.cpp
index 2996a505e..c38873faa 100644
--- a/fpdfsdk/fpdf_edittext.cpp
+++ b/fpdfsdk/fpdf_edittext.cpp
@@ -398,6 +398,11 @@ CPDF_Font* LoadCompositeFont(CPDF_Document* pDoc,
   return pDoc->LoadFont(fontDict);
 }
 
+CPDF_TextObject* CPDFTextObjectFromFPDFPageObject(FPDF_PAGEOBJECT page_object) {
+  auto* obj = CPDFPageObjectFromFPDFPageObject(page_object);
+  return obj ? obj->AsText() : nullptr;
+}
+
 }  // namespace
 
 FPDF_EXPORT FPDF_PAGEOBJECT FPDF_CALLCONV
@@ -475,6 +480,31 @@ FPDFText_SetFillColor(FPDF_PAGEOBJECT text_object,
   return FPDFPageObj_SetFillColor(text_object, R, G, B, A);
 }
 
+FPDF_EXPORT FPDF_BOOL FPDF_CALLCONV FPDFText_GetMatrix(FPDF_PAGEOBJECT text,
+                                                       double* a,
+                                                       double* b,
+                                                       double* c,
+                                                       double* d,
+                                                       double* e,
+                                                       double* f) {
+  if (!text || !a || !b || !c || !d || !e || !f)
+    return false;
+
+  CPDF_TextObject* pTextObj = CPDFTextObjectFromFPDFPageObject(text);
+  if (!pTextObj)
+    return false;
+
+  CFX_Matrix text_matrix = pTextObj->GetTextMatrix();
+  *a = text_matrix.a;
+  *b = text_matrix.b;
+  *c = text_matrix.c;
+  *d = text_matrix.d;
+  *e = text_matrix.e;
+  *f = text_matrix.f;
+
+  return true;
+}
+
 FPDF_EXPORT void FPDF_CALLCONV FPDFFont_Close(FPDF_FONT font) {
   CPDF_Font* pFont = CPDFFontFromFPDFFont(font);
   if (!pFont)
diff --git a/public/fpdf_edit.h b/public/fpdf_edit.h
index c0766a33b..5a2eeb449 100644
--- a/public/fpdf_edit.h
+++ b/public/fpdf_edit.h
@@ -954,6 +954,31 @@ FPDFText_SetFillColor(FPDF_PAGEOBJECT text_object,
                       unsigned int B,
                       unsigned int A);
 
+// Experimental API.
+// Get the transform matrix of a text object.
+//
+//   text - handle to a text.
+//   a    - matrix value.
+//   b    - matrix value.
+//   c    - matrix value.
+//   d    - matrix value.
+//   e    - matrix value.
+//   f    - matrix value.
+//
+// The matrix is composed as:
+//   |a c e|
+//   |b d f|
+// and used to scale, rotate, shear and translate the text.
+//
+// Returns TRUE on success.
+FPDF_EXPORT FPDF_BOOL FPDF_CALLCONV FPDFText_GetMatrix(FPDF_PAGEOBJECT text,
+                                                       double* a,
+                                                       double* b,
+                                                       double* c,
+                                                       double* d,
+                                                       double* e,
+                                                       double* f);
+
 // Close a loaded PDF font.
 //
 // font   - Handle to the loaded font.
-- 
2.16.4

